internal/pkg/image/driver/gocryptfsdriver.go

```diff
@@ -0,0 +1,464 @@
+	if params.Filesystem != "gocryptfs" {
+		return fmt.Errorf("filesystem type should be 'gocryptfs'")
+	}
+	g := &d.gocryptfsFeature
+	return g.Mount(params, mfunc, d.cmdPrefix)
+}
+
+func (d *gocryptfsDriver) Start(params *image.DriverParams, containerPid int) error {
+	if containerPid != 0 {
+		// Running in hybrid setuid-fakeroot mode
+		// Need any subcommand to first enter the container's
+		//  user namespace
+		nsenter, err := bin.FindBin("nsenter")
+		if err != nil {
+			return fmt.Errorf("failed to find nsenter: %v", err)
+		}
+		d.cmdPrefix = []string{
+			nsenter,
+			fmt.Sprintf("--user=/proc/%d/ns/user", containerPid),
+			"-F",
+		}
+	}
+	return nil
+}
+
+func (d *gocryptfsDriver) Stop(target string) error {
+	if err := d.gocryptfsFeature.stop(target, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *gocryptfsDriver) GetDriverName() string {
+	return gocryptfsDriverName
+}
+
+func CheckMountInfo(path string, ch chan error) {
+	iter := 0
+	for {
+		ok, err := mountPkg.Mounted(path)
+		if ok {
+			ch <- nil
+			break
+		}
+		if err != nil {
+			ch <- err
+			break
+		}
+		time.Sleep(1 * time.Second)
+		iter++
+		if iter >= 10 {
+			ch <- fmt.Errorf("timeout to check mount info for target path: %s, timeout limit: %d s", path, iter)
+			break
+		}
+	}
+}
+
+func GetGocryptfsPid(substr string) (int, error) {
+	d, err := os.Open("/proc")
+	if err != nil {
+		return -1, err
+	}
+	defer d.Close()
+	for {
+		names, err := d.Readdirnames(10)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return -1, err
+		}
+
+		for _, name := range names {
+			// We only care if the name starts with a numeric
+			if name[0] < '0' || name[0] > '9' {
+				continue
+			}
+
+			// From this point forward, any errors we just ignore, because
+			// it might simply be that the process doesn't exist anymore.
+			pid, err := strconv.ParseInt(name, 10, 0)
+			if err != nil {
+				continue
+			}
+
+			content, err := os.ReadFile(fmt.Sprintf("/proc/%d/cmdline", pid))
+			if err != nil {
+				return -1, err
+			}
+			if strings.Contains(string(content), substr) {
+				return int(pid), nil
+			}
+		}
+	}
+
+	return -1, fmt.Errorf("could not find any process containing the substr: %s", substr)
+}
+
+func KillProcess(pid int) error {
+	err := syscall.Kill(pid, syscall.SIGTERM)
+	if err != nil {
+		// child process not exit
+		if err.(syscall.Errno) == syscall.ECHILD {
+			return nil
+		}
+		return fmt.Errorf("send SIGTERM signal to process: %d encounters error: %v", pid, err)
+	}
+
+	var ws syscall.WaitStatus
+	wpid, err := syscall.Wait4(pid, &ws, syscall.WNOHANG, nil)
+	if err != nil {
+		// child process not exit
+		if err.(syscall.Errno) == syscall.ECHILD {
+			return nil
+		}
+		return fmt.Errorf("could not retrieve the process status for pid: %d, err: %v", pid, err)
+	}
+
+	if wpid != 0 {
+		return fmt.Errorf("process pid: %d exited with status: %v", pid, ws.ExitStatus())
+	}
+	return nil
+}
```

Comment:
This has a lot in common with the image driver's stop() function.  Let's share it.